### ISSUE TITLE
to_categorical should go before get_num_classes in metrics/functional/classification.py

### ISSUE THREAD

**Issue by pwwang** [PyTorchLightning/pytorch-lightning#2768]
https://github.com/PyTorchLightning/pytorch-lightning/blob/d18b9ef9d95ffd92591f767808cc497af5bd4e1c/pytorch_lightning/metrics/functional/classification.py#L174-L178

`to_categorical` should go before `get_num_classes`, since `get_num_classes` assumes `pred` has already been turned into categories (it is doing `int(pred.max().detach().item() + 1)`).

Warnings are raised now and then with current code, for example:
```
UserWarning: You have set 10 number of classes if different from predicted (xx) and target (10) number of classes
```

If logits are passed in directly for metrics.

**Comment by Borda** [comment]
@pwwang mind send a PR? 🐰
